flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/SqlTimestamp.java

```diff
@@ -0,0 +1,180 @@
+	 *                    A negative number is the number of milliseconds before
+	 *                    January 1, 1970, 00:00:00 GMT
+	 * @return an instance of {@code SqlTimestamp}
+	 */
+	public static SqlTimestamp fromEpochMillis(long millisecond) {
+		return new SqlTimestamp(millisecond, 0);
+	}
+
+	/**
+	 * Obtains an instance of {@code SqlTimestamp} from a millisecond and a nanoOfMillisecond.
+	 *
+	 * <p>This returns a {@code SqlTimestamp} with the specified millisecond and nanoOfMillisecond.
+	 *
+	 * @param millisecond the number of milliseconds since January 1, 1970, 00:00:00 GMT
+	 *                    A negative number is the number of milliseconds before
+	 *                    January 1, 1970, 00:00:00 GMT
+	 * @param nanoOfMillisecond the nanosecond within the millisecond, from 0 to 999,999
+	 * @return an instance of {@code SqlTimestamp}
+	 */
+	public static SqlTimestamp fromEpochMillis(long millisecond, int nanoOfMillisecond) {
+		return new SqlTimestamp(millisecond, nanoOfMillisecond);
+	}
+
+	/**
+	 * Convert this {@code SqlTimestmap} object to a {@link Timestamp}.
+	 *
+	 * @return An instance of {@link Timestamp}
+	 */
+	public Timestamp toTimestamp() {
+		return Timestamp.valueOf(toLocalDateTime());
+	}
+
+	/**
+	 * Obtains an instance of {@code SqlTimestamp} from an instance of {@link Timestamp}.
+	 *
+	 * <p>This returns a {@code SqlTimestamp} with the specified {@link Timestamp}.
+	 *
+	 * @param ts an instance of {@link Timestamp}
+	 * @return an instance of {@code SqlTimestamp}
+	 */
+	public static SqlTimestamp fromTimestamp(Timestamp ts) {
+		return fromLocalDateTime(ts.toLocalDateTime());
+	}
+
+	/**
+	 * Convert this {@code SqlTimestamp} object to a {@link LocalDateTime}.
+	 *
+	 * @return An instance of {@link LocalDateTime}
+	 */
+	public LocalDateTime toLocalDateTime() {
+		int date = (int) (millisecond / MILLIS_PER_DAY);
+		int time = (int) (millisecond % MILLIS_PER_DAY);
+		if (time < 0) {
+			--date;
+			time += MILLIS_PER_DAY;
+		}
+		long nanoOfDay = time * 1_000_000L + nanoOfMillisecond;
+		LocalDate localDate = LocalDate.ofEpochDay(date);
+		LocalTime localTime = LocalTime.ofNanoOfDay(nanoOfDay);
+		return LocalDateTime.of(localDate, localTime);
+	}
+
+	/**
+	 * Obtains an instance of {@code SqlTimestamp} from an instance of {@link LocalDateTime}.
+	 *
+	 * <p>This returns a {@code SqlTimestamp} with the specified {@link LocalDateTime}.
+	 *
+	 * @param dateTime an instance of {@link LocalDateTime}
+	 * @return an instance of {@code SqlTimestamp}
+	 */
+	public static SqlTimestamp fromLocalDateTime(LocalDateTime dateTime) {
+		long epochDay = dateTime.toLocalDate().toEpochDay();
+		long nanoOfDay = dateTime.toLocalTime().toNanoOfDay();
+
+		long millisecond = epochDay * MILLIS_PER_DAY + nanoOfDay / 1_000_000;
+		int nanoOfMillisecond = (int) (nanoOfDay % 1_000_000);
+
+		return new SqlTimestamp(millisecond, nanoOfMillisecond);
+	}
+
+	public static boolean isCompact(int precision) {
```

Comment:
add some comments on this? What does `compact` mean?

Reply:
Apache Calcite and Flink's planner use the number of milliseconds since epoch to represent a Timestamp type **compactly** if the number of digits of fractional seconds is between 0 - 3.

Reply:
Added comments